src/main/java/generator/ResponseHydrator.java

```diff
@@ -0,0 +1,165 @@
+                        childOverride = (Map<String, Object>) override.get(fieldName);
+                    }
+                    fieldResponse = hydrateSelectionRec(selectionType, field.getSelectionSet(), childOverride);
+                }
+
+                String selectionName = field.getAlias() == null ? field.getName() : field.getAlias();
+                children.put(selectionName, fieldResponse);
+            } else {
+                // TODO (othebe): Handle fragments.
+                throw new RuntimeException(String.format("Unhandled selection: %s", selection.toString()));
+            }
+        }
+
+        return ResponseNode.createBranchNode(children);
+    }
+
+    /**
+     * Provides the {@link graphql.schema.GraphQLOutputType} for a field in the graphQLType.
+     * @param graphQLType The object containing the field.
+     * @param fieldName The name of the field.
+     * @return {@link graphql.schema.GraphQLOutputType}
+     */
+    private GraphQLOutputType getOutputTypeForField(GraphQLType graphQLType, String fieldName) {
+        if (graphQLType instanceof GraphQLFieldsContainer) {
+            return ((GraphQLFieldsContainer) graphQLType).getFieldDefinition(fieldName).getType();
+        } else if (graphQLType instanceof GraphQLList) {
+            return getOutputTypeForField(((GraphQLList) graphQLType).getWrappedType(), fieldName);
+        } else {
+            throw new RuntimeException(String.format("Extracting fields from unhandled output type %s.%s", graphQLType.getName(), fieldName));
+        }
+    }
+
+    /**
+     * Uses IValueHydratorFactory to hydrate values in the response.
+     * @param graphQLType The value to hydrate.
+     * @param valueHydratorFactory Value hydrator provider.
+     * @return A hydrated value.
+     * TODO (othebe): Custom scalars.
+     */
+    private Object hydrateValue(GraphQLType graphQLType, IValueHydratorFactory valueHydratorFactory) {
+        GraphQLType resolvedType = graphQLType;
+
+        // Resolve types for wrapped data.
+        if (graphQLType instanceof GraphQLNonNull) {
+            resolvedType = ((GraphQLNonNull) graphQLType).getWrappedType();
+        }
+
+        // List.
+        // TODO (othebe): Add support for lists.
+        if (resolvedType instanceof GraphQLList) {
+            return hydrateValue(((GraphQLList) resolvedType).getWrappedType(), valueHydratorFactory);
+        }
+
+        // Enum.
+        else if (resolvedType instanceof GraphQLEnumType) {
+            return valueHydratorFactory.provideEnumHydrator().hydrateEnum(((GraphQLEnumType) resolvedType).getValues());
+        }
+
+        // Scalar.
+        // TODO (othebe): Incomplete.
+        else if (resolvedType instanceof GraphQLScalarType) {
+            IScalarHydrator scalarHydrator = valueHydratorFactory.provideScalarHydrator();
+
+            if (resolvedType.equals(Scalars.GraphQLString)) {
+                return scalarHydrator.hydrateString();
```

Comment:
I thought it would easier to have a map of <Type, Resolvers> [à-la graphql-tools](https://github.com/apollographql/graphql-tools/blob/master/src/mock.ts#L85-L90), so we don't have to create method signatures for all the types that we want to support (`hydrateInt()`, `hydrateX()`...)

Reply:
Yea, good point. We could define the hydrator/resolvers in a map, and only expose a single method hydrate/resolve(scalar) -> value. This would remove the if-block, and lookup would work for scalars since these are Singletons.